File: flowersec-go/controlplane/controlplane_test.go
package controlplane

import (
	"bytes"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"strings"
	"testing"
	"time"

	"github.com/floegence/flowersec/flowersec-go/v3/internal/artifactv3"
)

func TestIssuerCreatesOpaqueDirectArtifactAndBoundRuntimeAuthorization(t *testing.T) {
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x42}, 128)), time.Unix(1_800_000_000, 0))
	endpoints, err := NewEndpointSet(
		caEndpoint("websocket", "wss://edge.example/flowersec/v3/direct"),
		caEndpoint("raw-quic", "quic://edge.example"),
		caEndpoint("webtransport", "https://edge.example/flowersec/webtransport/v3/direct"),
	)
	if err != nil {
		t.Fatal(err)
	}
	issued, err := issuer.IssueDirect(DirectIssueOptions{
		Session:   SessionOptions{ChannelID: "channel-direct", ExpiresAt: time.Unix(1_800_000_060, 0)},
		Endpoints: endpoints, RendezvousGroupID: "group-direct", ListenerAudience: "audience-direct",
		UpstreamAddress: "127.0.0.1:9000",
		Metadata: ArtifactMetadata{
			Scopes:          []Scope{{Name: "redeven.proxy", Version: 1, Critical: true, Payload: json.RawMessage(`{"mode":"service_worker"}`)}},
			CorrelationTags: map[string]string{"trace": "trace-123"},
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := artifactv3.DecodeArtifactJSON(bytes.NewReader(issued.ArtifactJSON())); err != nil {
		t.Fatalf("public artifact parse failed: %v", err)
	}
	encodedRecord, err := issued.AuthorizationRecord().Encode()
	if err != nil {
		t.Fatal(err)
	}
	record, err := ParseAuthorizationRecord(encodedRecord)
	if err != nil {
		t.Fatal(err)
	}

	request := runtimeRequestForCandidate(t, record, "websocket")
	if request.LookupKey() != issued.LookupKey() {
		t.Fatalf("request lookup key = %q, want issued key %q", request.LookupKey(), issued.LookupKey())
	}
	response, err := AuthorizeRuntime(request, record, "lease-direct")
	if err != nil {
		t.Fatal(err)
	}
	var wire map[string]any
	if err := json.Unmarshal(response.JSON(), &wire); err != nil {
		t.Fatal(err)
	}
	if wire["decision"] != "allow" || wire["credential_id"] != issued.LookupKey() || wire["lease_id"] != "lease-direct" {
		t.Fatalf("unexpected authorization response: %#v", wire)
	}
	if _, present := wire["artifact"]; present || strings.Contains(string(response.JSON()), "e2ee_psk_b64u") {
		t.Fatalf("authorization response leaked artifact fields: %s", response.JSON())
	}
}

func TestRuntimeAuthorizationReturnsExactRetryAtArtifactExpiry(t *testing.T) {
	issuedAt := time.Unix(1_800_000_000, 0).UTC()
	expiresAt := issuedAt.Add(time.Minute)
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x5a}, 512)), issuedAt)
	directEndpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://edge.example/flowersec/v3/direct"))
	if err != nil {
		t.Fatal(err)
	}
	direct, err := issuer.IssueDirect(DirectIssueOptions{
		Session: SessionOptions{ChannelID: "expired-direct", ExpiresAt: expiresAt}, Endpoints: directEndpoints,
		RendezvousGroupID: "expired-direct-group", ListenerAudience: "expired-direct-audience",
		UpstreamAddress: "127.0.0.1:9000",
	})
	if err != nil {
		t.Fatal(err)
	}
	tunnelEndpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://edge.example/flowersec/v3/tunnel"))
	if err != nil {
		t.Fatal(err)
	}
	pair, err := issuer.IssueTunnelPair(TunnelIssueOptions{
		Session: SessionOptions{ChannelID: "expired-tunnel", ExpiresAt: expiresAt}, Endpoints: tunnelEndpoints,
		RendezvousGroupID: "expired-tunnel-group", ListenerAudience: "expired-tunnel-audience",
		FirstEndpointID: "expired-endpoint-a", SecondEndpointID: "expired-endpoint-b",
	})
	if err != nil {
		t.Fatal(err)
	}

	for _, test := range []struct {
		name      string
		issued    IssuedArtifact
		want      string
		authorize func(RuntimeAuthorizationRequest, AuthorizationRecord) ([]byte, error)
	}{
		{
			name: "direct", issued: direct,
			want: `{"decision":"retry","reason":"expired_artifact","credential_id":"","lease_id":"","expires_at":"0001-01-01T00:00:00Z","direct":null}`,
			authorize: func(request RuntimeAuthorizationRequest, record AuthorizationRecord) ([]byte, error) {
				response, err := authorizeRuntimeAt(request, record, "lease-expired-direct", expiresAt)
				return response.JSON(), err
			},
		},
		{
			name: "tunnel", issued: pair.First,
			want: `{"decision":"retry","reason":"expired_artifact","credential_id":"","lease_id":"","expires_at":"0001-01-01T00:00:00Z","expected_peer_endpoint_instance_id":"","allow_replacement":false}`,
			authorize: func(request RuntimeAuthorizationRequest, record AuthorizationRecord) ([]byte, error) {
				response, err := authorizeTunnelRuntimeAt(request, record, "lease-expired-tunnel", expiresAt)
				return response.JSON(), err
			},
		},
	} {
		t.Run(test.name, func(t *testing.T) {
			encodedRecord, err := test.issued.AuthorizationRecord().Encode()
			if err != nil {
				t.Fatal(err)
			}
			record, err := ParseAuthorizationRecord(encodedRecord)
			if err != nil {
				t.Fatal(err)
			}
			requestBody := runtimeRequestBody(t, record, "websocket", "")
			request, err := ParseRuntimeAuthorizationRequest(requestBody)
			if err != nil {
				t.Fatal(err)
			}
			encoded, err := test.authorize(request, record)
			if err != nil {
				t.Fatalf("expired artifact returned an input error: %v", err)
			}
			if string(encoded) != test.want {
				t.Fatalf("expired artifact response = %s, want %s", encoded, test.want)
			}
		})
	}
}

func TestProductionIssuerMatchesSharedAdmissionHashVector(t *testing.T) {
	var fixture struct {
		ArtifactJSON              string `json:"artifact_json"`
		ChosenCandidateID         string `json:"chosen_candidate_id"`
		FSB3Hex                   string `json:"fsb3_hex"`
		AdmissionBindingHex       string `json:"admission_binding_hex"`
		AcceptorAdmissionsHashHex string `json:"acceptor_admissions_hash_hex"`
	}
	raw, err := os.ReadFile("../../testdata/transport_v3/go_issuer_admission_vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(raw, &fixture); err != nil {
		t.Fatal(err)
	}
	now := time.Unix(1_800_000_000, 0).UTC()
	endpoints, err := NewEndpointSet(caEndpoint("issuer-ws", "wss://issuer.example/flowersec/v3/direct"))
	if err != nil {
		t.Fatal(err)
	}
	issued, err := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x42}, 128)), now).IssueDirect(DirectIssueOptions{
		Session:   SessionOptions{ChannelID: "issuer-shared", ExpiresAt: now.Add(time.Minute)},
		Endpoints: endpoints, RendezvousGroupID: "issuer-group", ListenerAudience: "issuer-audience",
		UpstreamAddress: "127.0.0.1:9000",
	})
	if err != nil {
		t.Fatal(err)
	}
	if string(issued.ArtifactJSON()) != fixture.ArtifactJSON {
		t.Fatal("Go production issuer artifact differs from shared fixture")
	}
	artifact, err := artifactv3.DecodeArtifactJSON(bytes.NewReader(issued.ArtifactJSON()))
	if err != nil {
		t.Fatal(err)
	}
	request, err := artifactv3.BuildRequest(*artifact, fixture.ChosenCandidateID)
	if err != nil {
		t.Fatal(err)
	}
	frame, err := artifactv3.MarshalRequest(request)
	if err != nil {
		t.Fatal(err)
	}
	actualHex := hex.EncodeToString(frame)
	if actualHex != fixture.FSB3Hex {
		t.Fatalf("Go production FSB3 differs: got %d hex chars want %d", len(actualHex), len(fixture.FSB3Hex))
	}
	binding := artifactv3.AdmissionBinding(frame)
	if hex.EncodeToString(binding[:]) != fixture.AdmissionBindingHex {
		t.Fatal("Go admission binding differs from shared fixture")
	}
	acceptor, err := artifactv3.AcceptorAdmissionsHash([][]byte{frame})
	if err != nil {
		t.Fatal(err)
	}
	if hex.EncodeToString(acceptor[:]) != fixture.AcceptorAdmissionsHashHex {
		t.Fatal("Go acceptor admissions hash differs from shared fixture")
	}
}

func TestIssuerRoundsSubsecondExpiryUpToWireSecond(t *testing.T) {
	now := time.Unix(1_800_000_000, 500_000_000)
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x61}, 128)), now)
	endpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://edge.example/flowersec/v3/direct"))
	if err != nil {
		t.Fatal(err)
	}
	issued, err := issuer.IssueDirect(DirectIssueOptions{
		Session:   SessionOptions{ChannelID: "subsecond", ExpiresAt: now.Add(100 * time.Millisecond)},
		Endpoints: endpoints, RendezvousGroupID: "group", ListenerAudience: "audience",
		UpstreamAddress: "127.0.0.1:9000",
	})
	if err != nil {
		t.Fatal(err)
	}
	artifact, err := artifactv3.DecodeArtifactJSON(bytes.NewReader(issued.ArtifactJSON()))
	if err != nil {
		t.Fatal(err)
	}
	if got, want := artifact.Session.InitExpireAtUnixSeconds, now.Unix()+1; got != want {
		t.Fatalf("wire expiry = %d, want rounded-up second %d", got, want)
	}
}

func TestEndpointSetRejectsPlainWebSocketIncludingLoopback(t *testing.T) {
	for _, rawURL := range []string{
		"ws://127.0.0.1:23998/flowersec/v3/direct",
		"ws://edge.example:23998/flowersec/v3/direct",
	} {
		_, err := NewEndpointSet(caEndpoint("websocket", rawURL))
		var controlErr *ControlPlaneError
		if !errors.As(err, &controlErr) || controlErr.Code() != InvalidEndpointURL || controlErr.FieldPath() != "endpoints[0].url" {
			t.Fatalf("plain WebSocket error = %#v, want invalid_endpoint_url", err)
		}
	}
}

func TestControlPlaneBoundariesFailClosedAndRedactOpaqueValues(t *testing.T) {
	now := time.Unix(1_800_000_000, 0)
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x31}, 128)), now)
	directEndpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://edge.example/flowersec/v3/direct"))
	if err != nil {
		t.Fatal(err)
	}
	if _, err := issuer.IssueDirect(DirectIssueOptions{
		Session: SessionOptions{ChannelID: "expired", ExpiresAt: now}, Endpoints: directEndpoints,
		RendezvousGroupID: "group", ListenerAudience: "audience", UpstreamAddress: "127.0.0.1:9000",
	}); err == nil {
		t.Fatal("expired issuance unexpectedly succeeded")
	}
	if _, err := issuer.IssueDirect(DirectIssueOptions{
		Session: SessionOptions{ChannelID: "bad-upstream", ExpiresAt: now.Add(time.Minute)}, Endpoints: directEndpoints,
		RendezvousGroupID: "group", ListenerAudience: "audience", UpstreamAddress: "missing-port",
	}); err == nil {
		t.Fatal("invalid direct upstream unexpectedly succeeded")
	}

	issued, err := issuer.IssueDirect(DirectIssueOptions{
		Session: SessionOptions{ChannelID: "redacted", ExpiresAt: now.Add(time.Minute)}, Endpoints: directEndpoints,
		RendezvousGroupID: "group", ListenerAudience: "audience", UpstreamAddress: "127.0.0.1:9000",
	})
	if err != nil {
		t.Fatal(err)
	}
	record := issued.AuthorizationRecord()
	request := runtimeRequestForCandidate(t, record, "websocket")
	response, err := AuthorizeRuntime(request, record, "lease-redacted")
	if err != nil {
		t.Fatal(err)
	}
	for name, value := range map[string]any{
		"issued": issued, "record": record, "request": request, "response": response,
	} {
		encoded, err := json.Marshal(value)
		if err != nil || string(encoded) != "{}" {
			t.Fatalf("%s generic JSON = %s, %v", name, encoded, err)
		}
		formatted := fmt.Sprintf("%+v", value)
		if strings.Contains(formatted, "edge.example") || strings.Contains(formatted, "127.0.0.1") || strings.Contains(formatted, "channel") {
			t.Fatalf("%s formatting leaked sensitive data: %s", name, formatted)
		}
	}

	encodedRecord, err := record.Encode()
	if err != nil {
		t.Fatal(err)
	}
	var recordWire map[string]any
	if err := json.Unmarshal(encodedRecord, &recordWire); err != nil {
		t.Fatal(err)
	}
	recordWire["lookup_key"] = strings.Repeat("a", 43)
	tampered, _ := json.Marshal(recordWire)
	if _, err := ParseAuthorizationRecord(tampered); err == nil {
		t.Fatal("tampered record lookup key unexpectedly succeeded")
	}

	requestBody := runtimeRequestBody(t, record, "websocket", "websocket")
	var requestWire map[string]any
	if err := json.Unmarshal(requestBody, &requestWire); err != nil {
		t.Fatal(err)
	}
	requestWire["unknown"] = true
	unknown, _ := json.Marshal(requestWire)
	if _, err := ParseRuntimeAuthorizationRequest(unknown); err == nil {
		t.Fatal("unknown runtime request field unexpectedly succeeded")
	}
	requestWire = map[string]any{}
	if err := json.Unmarshal(requestBody, &requestWire); err != nil {
		t.Fatal(err)
	}
	requestWire["carrier"] = "raw_quic"
	drifted, _ := json.Marshal(requestWire)
	if _, err := ParseRuntimeAuthorizationRequest(drifted); err == nil {
		t.Fatal("runtime carrier drift unexpectedly succeeded")
	}
}

func TestIssuerRedactsRandomSourceFailure(t *testing.T) {
	issuer := newIssuerForTest(errorReader{}, time.Unix(1_800_000_000, 0))
	endpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://edge.example/flowersec/v3/direct"))
	if err != nil {
		t.Fatal(err)
	}
	_, err = issuer.IssueDirect(DirectIssueOptions{
		Session:   SessionOptions{ChannelID: "channel", ExpiresAt: time.Unix(1_800_000_060, 0)},
		Endpoints: endpoints, RendezvousGroupID: "group", ListenerAudience: "audience",
		UpstreamAddress: "127.0.0.1:9000",
	})
	if !errors.Is(err, ErrIssuanceFailed) || err.Error() != ErrIssuanceFailed.Error() {
		t.Fatalf("random source error = %v, want stable redacted issuance failure", err)
	}
}

func TestIssuerRedactsClockAndProviderFailures(t *testing.T) {
	for _, test := range []struct {
		name   string
		issuer *Issuer
	}{
		{name: "missing clock", issuer: &Issuer{random: bytes.NewReader(make([]byte, 64))}},
		{name: "zero clock", issuer: &Issuer{
			random: bytes.NewReader(make([]byte, 64)),
			now:    func() time.Time { return time.Time{} },
		}},
		{name: "missing random provider", issuer: &Issuer{now: time.Now}},
	} {
		t.Run(test.name, func(t *testing.T) {
			_, err := test.issuer.IssueDirect(DirectIssueOptions{})
			if !errors.Is(err, ErrIssuanceFailed) || errors.Is(err, ErrInvalidControlPlaneInput) ||
				err.Error() != ErrIssuanceFailed.Error() {
				t.Fatalf("provider failure = %v, want stable redacted issuance failure", err)
			}
		})
	}
}

func TestRejectRuntimeBuildsOnlyValidatedRejectAndRetryResponses(t *testing.T) {
	for _, test := range []struct {
		retryable bool
		decision  string
	}{{false, "reject"}, {true, "retry"}} {
		response, err := RejectRuntime("permission_denied", test.retryable)
		if err != nil {
			t.Fatal(err)
		}
		var wire map[string]any
		if err := json.Unmarshal(response.JSON(), &wire); err != nil {
			t.Fatal(err)
		}
		if wire["decision"] != test.decision || wire["reason"] != "permission_denied" {
			t.Fatalf("rejection response = %#v", wire)
		}
	}
	if _, err := RejectRuntime("Secret Detail", false); err == nil {
		t.Fatal("invalid rejection reason unexpectedly succeeded")
	}
	if _, err := RejectRuntime(artifactv3.ReasonExpiredArtifact, false); !errors.Is(err, ErrInvalidControlPlaneInput) {
		t.Fatalf("expired artifact reject error = %v, want invalid control-plane input", err)
	}
	if _, err := RejectTunnelRuntime(artifactv3.ReasonExpiredArtifact, false); !errors.Is(err, ErrInvalidControlPlaneInput) {
		t.Fatalf("expired tunnel artifact reject error = %v, want invalid control-plane input", err)
	}
}

type errorReader struct{}

func (errorReader) Read([]byte) (int, error) { return 0, errors.New("secret provider detail") }

func TestIssuerCreatesTunnelPairAndRejectsCrossRecordAuthorization(t *testing.T) {
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x24}, 256)), time.Unix(1_800_000_000, 0))
	endpoints, err := NewEndpointSet(
		caEndpoint("websocket", "wss://tunnel.example/flowersec/v3/tunnel"),
		caEndpoint("raw-quic", "quic://tunnel.example"),
		caEndpoint("webtransport", "https://tunnel.example/flowersec/webtransport/v3/tunnel"),
	)
	if err != nil {
		t.Fatal(err)
	}
	pair, err := issuer.IssueTunnelPair(TunnelIssueOptions{
		Session:   SessionOptions{ChannelID: "channel-tunnel", ExpiresAt: time.Unix(1_800_000_060, 0)},
		Endpoints: endpoints, RendezvousGroupID: "group-tunnel", ListenerAudience: "audience-tunnel",
		FirstEndpointID: "endpoint-a", SecondEndpointID: "endpoint-b", AllowReplacement: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	firstRecord := pair.First.AuthorizationRecord()
	secondRecord := pair.Second.AuthorizationRecord()
	firstRequest := runtimeRequestForCandidate(t, firstRecord, "raw-quic")
	if _, err := AuthorizeRuntime(firstRequest, secondRecord, "lease-cross"); err == nil {
		t.Fatal("cross-record authorization unexpectedly succeeded")
	}
	if _, err := AuthorizeRuntime(firstRequest, firstRecord, "lease-first"); err == nil {
		t.Fatal("direct runtime authorization accepted a tunnel request")
	}
}

func TestTunnelRuntimeAuthorizationNeverContainsSessionSecrets(t *testing.T) {
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x25}, 256)), time.Unix(1_800_000_000, 0))
	endpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://tunnel.example/flowersec/v3/tunnel"))
	if err != nil {
		t.Fatal(err)
	}
	pair, err := issuer.IssueTunnelPair(TunnelIssueOptions{
		Session:   SessionOptions{ChannelID: "opaque-relay", ExpiresAt: time.Unix(1_800_000_060, 0)},
		Endpoints: endpoints, RendezvousGroupID: "opaque-group", ListenerAudience: "opaque-audience",
		FirstEndpointID: "endpoint-a", SecondEndpointID: "endpoint-b",
	})
	if err != nil {
		t.Fatal(err)
	}
	record := pair.First.AuthorizationRecord()
	request := runtimeRequestForCandidate(t, record, "websocket")
	response, err := AuthorizeTunnelRuntime(request, record, "lease-opaque")
	if err != nil {
		t.Fatal(err)
	}
	var wire map[string]any
	if err := json.Unmarshal(response.JSON(), &wire); err != nil {
		t.Fatal(err)
	}
	for _, forbidden := range []string{"session", "direct", "e2ee_psk_base64url", "allowed_suites", "default_suite"} {
		if _, exists := wire[forbidden]; exists {
			t.Fatalf("tunnel runtime authorization exposes forbidden field %q", forbidden)
		}
		if bytes.Contains(response.JSON(), []byte(forbidden)) {
			t.Fatalf("tunnel runtime authorization contains forbidden token %q", forbidden)
		}
	}
}

func TestAllowTunnelRuntimeBuildsSecretFreeResponseAfterExternalAuthorization(t *testing.T) {
	issuer := newIssuerForTest(bytes.NewReader(bytes.Repeat([]byte{0x26}, 256)), time.Unix(1_800_000_000, 0))
	endpoints, err := NewEndpointSet(caEndpoint("websocket", "wss://tunnel.example/flowersec/v3/tunnel"))
	if err != nil {
		t.Fatal(err)
	}
	pair, err := issuer.IssueTunnelPair(TunnelIssueOptions{
		Session:   SessionOptions{ChannelID: "external-authorization", ExpiresAt: time.Unix(1_800_000_060, 0)},
		Endpoints: endpoints, RendezvousGroupID: "external-group", ListenerAudience: "external-audience",
		FirstEndpointID: "endpoint-a", SecondEndpointID: "endpoint-b", AllowReplacement: true,
	})
	if err != nil {
		t.Fatal(err)
	}
	request := runtimeRequestForCandidate(t, pair.First.AuthorizationRecord(), "websocket")
	response, err := AllowTunnelRuntime(request, "lease-external", time.Now().Add(time.Minute), "endpoint-b", true)
	if err != nil {
		t.Fatal(err)
	}
	var wire map[string]any
	if err := json.Unmarshal(response.JSON(), &wire); err != nil {
		t.Fatal(err)
	}
	if wire["decision"] != "allow" || wire["credential_id"] != request.LookupKey() || wire["expected_peer_endpoint_instance_id"] != "endpoint-b" {
		t.Fatalf("unexpected tunnel authorization response: %#v", wire)
	}
	for _, forbidden := range []string{"session", "artifact", "psk", "secret"} {
		if bytes.Contains(bytes.ToLower(response.JSON()), []byte(forbidden)) {
			t.Fatalf("secret-free response contains forbidden token %q", forbidden)
		}
	}

	direct, err := issuer.IssueDirect(DirectIssueOptions{
		Session:   SessionOptions{ChannelID: "direct-external", ExpiresAt: time.Unix(1_800_000_060, 0)},
		Endpoints: endpoints, RendezvousGroupID: "direct-group", ListenerAudience: "direct-audience",
		UpstreamAddress: "127.0.0.1:9000",
	})
	if err == nil {
		directRequest := runtimeRequestForCandidate(t, direct.AuthorizationRecord(), "websocket")
		if _, allowErr := AllowTunnelRuntime(directRequest, "lease-direct", time.Now().Add(time.Minute), "endpoint-b", false); allowErr == nil {
			t.Fatal("secret-free tunnel allow accepted a direct request")
		}
	}
	if _, err := AllowTunnelRuntime(request, "lease-external", time.Now().Add(-time.Second), "endpoint-b", false); err == nil {
		t.Fatal("secret-free tunnel allow accepted an expired lease")
	}
}

func runtimeRequestForCandidate(t *testing.T, record AuthorizationRecord, candidateID string) RuntimeAuthorizationRequest {
	t.Helper()
	body := runtimeRequestBody(t, record, candidateID, "")
	parsed, err := ParseRuntimeAuthorizationRequest(body)
	if err != nil {
		t.Fatal(err)
	}
	return parsed
}

func runtimeRequestBody(t *testing.T, record AuthorizationRecord, candidateID string, observedCarrier string) []byte {
	t.Helper()
	request, err := artifactv3.BuildRequest(*record.artifact, candidateID)
	if err != nil {
		t.Fatal(err)
	}
	raw, err := artifactv3.MarshalRequest(request)
	if err != nil {
		t.Fatal(err)
	}
	carrier := ""
	for _, candidate := range request.Candidates {
		if candidate.ID == candidateID {
			carrier = string(candidate.Carrier)
		}
	}
	if observedCarrier != "" {
		carrier = observedCarrier
	}
	body, err := json.Marshal(map[string]string{
		"fsb3_base64url": base64.RawURLEncoding.EncodeToString(raw),
		"carrier":        carrier,
		"remote_address": "198.51.100.10:50000",
	})
	if err != nil {
		t.Fatal(err)
	}
	return body
}

func caEndpoint(id, rawURL string) EndpointConfig {
	return EndpointConfig{ID: id, URL: rawURL, TLS: CAPolicy()}
}

File: flowersec-ts/src/node/controlplane.test.ts
import { readFileSync } from "node:fs";

import { describe, expect, test } from "vitest";
import {
  buildFSB2RequestV2,
  decodeArtifactV2JSON,
  decodeFSB2RequestV2,
  encodeFSB2RequestV2,
} from "../v2/artifact.js";
import { base64urlEncode } from "../utils/base64url.js";

import {
  AuthorizationRecord,
  Issuer,
  RuntimeAuthorizationRequest,
  authorizeTunnelRuntime,
  authorizeRuntime,
  createEndpointSet,
  rejectRuntime,
} from "./v2.js";
import type { AuthorizationDecision } from "./acceptor.js";

describe("Node control-plane public contract", () => {
  test("issues, persists, parses, and authorizes a direct artifact", () => {
    const now = Math.floor(Date.now() / 1000);
    const issuer = new Issuer();
    const endpoints = createEndpointSet(
      "wss://edge.example/flowersec/v2/direct",
      "quic://edge.example",
      "https://edge.example/flowersec/webtransport/v2/direct",
    );
    const issued = issuer.issueDirect({
      session: { channelId: "direct", expiresAtUnixSeconds: now + 60 },
      endpoints,
      rendezvousGroupId: "group",
      listenerAudience: "audience",
      upstreamAddress: "127.0.0.1:9000",
    });
    const artifact = issued.artifactJSON();
    expect(artifact).toBeInstanceOf(Uint8Array);
    const record = AuthorizationRecord.parse(issued.authorizationRecord().encode());
    expect(record.lookupKey()).toBe(issued.lookupKey());

    const decodedArtifact = decodeArtifactV2JSON(artifact);
    const candidate = decodedArtifact.path.candidates[0]!;
    const request = RuntimeAuthorizationRequest.parse(JSON.stringify({
      fsb2_base64url: base64urlEncode(encodeFSB2RequestV2(buildFSB2RequestV2(decodedArtifact, candidate.id))),
      carrier: candidate.carrier,
      remote_address: "127.0.0.1:23998",
    }));
    const response = authorizeRuntime(request, record, "lease-direct", now);
    const decision: AuthorizationDecision = response;
    expect(decision.decision).toBe("allow");
    expect(response.leaseId).toBe("lease-direct");
    expect(response.artifact).toBeDefined();
    expect(JSON.parse(new TextDecoder().decode(response.json()))).toMatchObject({
      decision: "allow",
      credential_id: issued.lookupKey(),
      lease_id: "lease-direct",
      direct: { upstream: { network: "tcp", address: "127.0.0.1:9000" } },
    });
  });

  test("issues a tunnel pair and rejects a cross-record request", () => {
    const now = Math.floor(Date.now() / 1000);
    const issuer = new Issuer();
    const pair = issuer.issueTunnelPair({
      session: { channelId: "tunnel", expiresAtUnixSeconds: now + 60 },
      endpoints: createEndpointSet("wss://edge.example/flowersec/v2/tunnel"),
      rendezvousGroupId: "group",
      listenerAudience: "audience",
      firstEndpointId: "endpoint-a",
      secondEndpointId: "endpoint-b",
      allowReplacement: true,
    });
    const firstRecord = pair.first.authorizationRecord();
    const secondRecord = pair.second.authorizationRecord();
    const artifact = decodeArtifactV2JSON(pair.first.artifactJSON());
    const candidate = artifact.path.candidates[0]!;
    const request = RuntimeAuthorizationRequest.parse(JSON.stringify({
      fsb2_base64url: base64urlEncode(encodeFSB2RequestV2(buildFSB2RequestV2(artifact, candidate.id))),
      carrier: candidate.carrier,
      remote_address: "127.0.0.1:23998",
    }));
    expect(() => authorizeRuntime(request, secondRecord, "lease-cross", now)).toThrow(/invalid/u);
    expect(() => authorizeRuntime(request, firstRecord, "lease-first", now)).toThrow(/invalid/u);
    const response = JSON.parse(new TextDecoder().decode(authorizeTunnelRuntime(request, firstRecord, "lease-first", now).json())) as Record<string, unknown>;
    expect(response).toMatchObject({
      decision: "allow",
      expected_peer_endpoint_instance_id: "endpoint-b",
      allow_replacement: true,
    });
    expect(Object.keys(response).sort()).toEqual([
      "allow_replacement",
      "credential_id",
      "decision",
      "expected_peer_endpoint_instance_id",
      "expires_at",
      "lease_id",
    ]);
    expect(Object.keys(response)).not.toEqual(expect.arrayContaining([
      expect.stringMatching(/session|psk|suite|secret/iu),
    ]));
    const secretMarkers = [
      artifact.session.e2ee_psk_b64u,
      ...(artifact.path.kind === "tunnel" ? [artifact.path.token] : []),
    ];
    for (const marker of secretMarkers) {
      expect(Object.values(response)).not.toContain(marker);
    }
  });

  test("compares bound admission requests with the Node constant-time primitive", () => {
    const now = Math.floor(Date.now() / 1000);
    const issued = new Issuer().issueDirect({
      session: { channelId: "constant-time", expiresAtUnixSeconds: now + 60 },
      endpoints: createEndpointSet("wss://edge.example/flowersec/v2/direct"),
      rendezvousGroupId: "constant-time-group",
      listenerAudience: "constant-time-audience",
      upstreamAddress: "127.0.0.1:9000",
    });
    const artifact = decodeArtifactV2JSON(issued.artifactJSON());
    const candidate = artifact.path.candidates[0]!;
    const raw = encodeFSB2RequestV2(buildFSB2RequestV2(artifact, candidate.id));
    const decoded = decodeFSB2RequestV2(raw);
    const authorize = (candidateRaw: Uint8Array) => authorizeRuntime(
      RuntimeAuthorizationRequest.fromDecoded({ ...decoded, raw: candidateRaw }, candidate.carrier),
      issued.authorizationRecord(),
      "constant-time-lease",
      now,
    );

    expect(() => authorize(raw.slice())).not.toThrow();
    const first = raw.slice();
    first[0] = (first[0] ?? 0) ^ 1;
    expect(() => authorize(first)).toThrow();
    const last = raw.slice();
    last[last.length - 1] = (last[last.length - 1] ?? 0) ^ 1;
    expect(() => authorize(last)).toThrow();
    expect(() => authorize(raw.subarray(0, raw.length - 1))).toThrow();

    const source = readFileSync(new URL("./controlplane.ts", import.meta.url), "utf8");
    expect(source).toMatch(/import\s*\{[^}]*timingSafeEqual[^}]*\}\s*from\s*["']node:crypto["']/su);
    expect(source).toMatch(/return timingSafeEqual\(a, b\)/u);
  });

  test("strictly validates runtime requests and bounded reject/retry responses", () => {
    expect(() => RuntimeAuthorizationRequest.parse(JSON.stringify({ fsb2_base64url: "bad", carrier: "websocket", remote_address: "127.0.0.1" }))).toThrow();
    expect(JSON.parse(new TextDecoder().decode(rejectRuntime("permission_denied", false).json()))).toEqual({ decision: "reject", reason: "permission_denied" });
    expect(JSON.parse(new TextDecoder().decode(rejectRuntime("busy", true).json()))).toEqual({ decision: "retry", reason: "busy" });
    expect(() => rejectRuntime("Secret Detail", false)).toThrow();
  });

  test("returns typed tunnel decisions without requiring response JSON parsing", () => {
    const now = Math.floor(Date.now() / 1000);
    const issuer = new Issuer();
    const pair = issuer.issueTunnelPair({
      session: { channelId: "typed-tunnel", expiresAtUnixSeconds: now + 60 },
      endpoints: createEndpointSet("wss://edge.example/flowersec/v2/tunnel"),
      rendezvousGroupId: "group",
      listenerAudience: "audience",
      firstEndpointId: "endpoint-a",
      secondEndpointId: "endpoint-b",
    });
    const record = pair.first.authorizationRecord();
    const artifact = decodeArtifactV2JSON(pair.first.artifactJSON());
    const candidate = artifact.path.candidates[0]!;
    const request = RuntimeAuthorizationRequest.parse(JSON.stringify({
      fsb2_base64url: base64urlEncode(encodeFSB2RequestV2(buildFSB2RequestV2(artifact, candidate.id))),
      carrier: candidate.carrier,
      remote_address: "127.0.0.1:23998",
    }));
    const response = authorizeTunnelRuntime(request, record, "lease-tunnel", now);
    expect(response.decision).toBe("allow");
    expect(response.leaseId).toBe("lease-tunnel");
    expect(response.expectedPeerEndpointInstanceId).toBe("endpoint-b");
    expect(response.reason).toBeUndefined();
  });
});
